File: examples/contrib/max_flow_taha.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Google.OrTools.ConstraintSolver;

public class MaxFlowTaha
{
    /**
     *
     * Max flow problem.
     *
     * From Taha "Introduction to Operations Research", Example 6.4-2
     *
     * Translated from the AMPL code at
     * http://taha.ineg.uark.edu/maxflo.txt
     *
     * Also see http://www.hakank.org/or-tools/max_flow_taha.py
     *
     */
    private static void Solve()
    {
        Solver solver = new Solver("MaxFlowTaha");

        //
        // Data
        //
        int n = 5;
        int start = 0;
        int end = n - 1;

        IEnumerable<int> NODES = Enumerable.Range(0, n);

        // cost matrix
        int[,] c = {
            { 0, 20, 30, 10, 0 }, { 0, 0, 40, 0, 30 }, { 0, 0, 0, 10, 20 }, { 0, 0, 5, 0, 20 }, { 0, 0, 0, 0, 0 }
        };

        //
        // Decision variables
        //
        IntVar[,] x = new IntVar[n, n];
        foreach (int i in NODES)
        {
            foreach (int j in NODES)
            {
                x[i, j] = solver.MakeIntVar(0, c[i, j], "x");
            }
        }

        IntVar[] x_flat = x.Flatten();

        IntVar[] out_flow = solver.MakeIntVarArray(n, 0, 1000, "out_flow");
        IntVar[] in_flow = solver.MakeIntVarArray(n, 0, 1000, "in_flow");
        IntVar total = solver.MakeIntVar(0, 10000, "total");

        //
        // Constraints
        //
    solver.Add((from j in NODES where c[start, j] > 0 select x[start, j])
                   .ToArray()
                   .Sum() == total);

    foreach (int i in NODES)
    {
      var in_flow_sum = (from j in NODES where c[j, i] > 0 select x[j, i]);
      if (in_flow_sum.Count() > 0)
      {
          solver.Add(in_flow_sum.ToArray().Sum() == in_flow[i]);
      }

      var out_flow_sum = (from j in NODES where c[i, j] > 0 select x[i, j]);
      if (out_flow_sum.Count() > 0)
      {
          solver.Add(out_flow_sum.ToArray().Sum() == out_flow[i]);
      }
    }

    // in_flow == out_flow
    foreach (int i in NODES)
    {
        if (i != start && i != end)
        {
            solver.Add(out_flow[i] == in_flow[i]);
        }
    }

    var s1 = (from i in NODES where c[i, start] > 0 select x[i, start]);
    if (s1.Count() > 0)
    {
        solver.Add(s1.ToArray().Sum() == 0);
    }

    var s2 = (from j in NODES where c[end, j] > 0 select x[end, j]);
    if (s2.Count() > 0)
    {
        solver.Add(s2.ToArray().Sum() == 0);
    }

    //
    // Objective
    //
    OptimizeVar obj = total.Maximize(1);

    //
    // Search
    //
    DecisionBuilder db = solver.MakePhase(x_flat.Concat(in_flow).Concat(out_flow).ToArray(), Solver.INT_VAR_DEFAULT,
                                          Solver.ASSIGN_MAX_VALUE);

    solver.NewSearch(db, obj);
    while (solver.NextSolution())
    {
        Console.WriteLine("total: {0}", total.Value());
        Console.Write("in_flow : ");
        foreach (int i in NODES)
        {
            Console.Write(in_flow[i].Value() + " ");
        }
        Console.Write("\nout_flow: ");
        foreach (int i in NODES)
        {
            Console.Write(out_flow[i].Value() + " ");
        }
        Console.WriteLine();
        foreach (int i in NODES)
        {
            foreach (int j in NODES)
            {
                Console.Write("{0,2} ", x[i, j].Value());
            }
            Console.WriteLine();
        }
        Console.WriteLine();
    }

    Console.WriteLine("\nSolutions: {0}", solver.Solutions());
    Console.WriteLine("WallTime: {0}ms", solver.WallTime());
    Console.WriteLine("Failures: {0}", solver.Failures());
    Console.WriteLine("Branches: {0} ", solver.Branches());

    solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        Solve();
    }
}

File: examples/contrib/max_flow_taha.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Max flow problem in Google CP Solver.

From Taha 'Introduction to Operations Research', Example 6.4-2

Translated from the AMPL code at
http://taha.ineg.uark.edu/maxflo.txt

Compare with the following model:
* MiniZinc: http://www.hakank.org/minizinc/max_flow_taha.mzn

This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

from ortools.constraint_solver.python import constraint_solver as cp


def main():

    # Create the solver.
    solver = cp.Solver("Max flow problem, Taha")

    #
    # data
    #
    n = 5
    start = 0
    end = n - 1

    nodes = list(range(n))

    # cost matrix
    c = [
        [0, 20, 30, 10, 0],
        [0, 0, 40, 0, 30],
        [0, 0, 0, 10, 20],
        [0, 0, 5, 0, 20],
        [0, 0, 0, 0, 0],
    ]

    #
    # declare variables
    #
    x = {}
    for i in nodes:
        for j in nodes:
            x[i, j] = solver.new_int_var(0, c[i][j], "x[%i,%i]" % (i, j))

    x_flat = [x[i, j] for i in nodes for j in nodes]
    out_flow = [solver.new_int_var(0, 10000, "out_flow[%i]" % i) for i in nodes]
    in_flow = [solver.new_int_var(0, 10000, "in_flow[%i]" % i) for i in nodes]

    total = solver.new_int_var(0, 10000, "z")

    #
    # constraints
    #
    cost_sum = solver.sum([x[start, j] for j in nodes if c[start][j] > 0])
    solver.add(total == cost_sum)

    for i in nodes:
        in_flow_sum = solver.sum([x[j, i] for j in nodes if c[j][i] > 0])
        solver.add(in_flow[i] == in_flow_sum)

        out_flow_sum = solver.sum([x[i, j] for j in nodes if c[i][j] > 0])
        solver.add(out_flow[i] == out_flow_sum)

    # in_flow == out_flow
    for i in nodes:
        if i != start and i != end:
            solver.add(out_flow[i] - in_flow[i] == 0)

    s1 = [x[i, start] for i in nodes if c[i][start] > 0]
    if len(s1) > 0:
        solver.add(solver.sum([x[i, start] for i in nodes if c[i][start] > 0] == 0))

    s2 = [x[end, j] for j in nodes if c[end][j] > 0]
    if len(s2) > 0:
        solver.add(solver.sum([x[end, j] for j in nodes if c[end][j] > 0]) == 0)

    # objective: maximize total cost
    objective = solver.maximize(total, 1)

    #
    # solution and search
    #
    db = solver.phase(
        x_flat,
        cp.IntVarStrategy.INT_VAR_DEFAULT,
        cp.IntValueStrategy.ASSIGN_MAX_VALUE,
    )

    solver.new_search(db, [objective])
    num_solutions = 0
    while solver.next_solution():
        num_solutions += 1
        print("total:", total.value())
        print("in_flow:", [in_flow[i].value() for i in nodes])
        print("out_flow:", [out_flow[i].value() for i in nodes])
        for i in nodes:
            for j in nodes:
                print("%2i" % x[i, j].value(), end=" ")
            print()
        print()

    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms, "ms")


if __name__ == "__main__":
    main()
